cpp/include/cuml/cluster/hdbscan.hpp

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright (c) 2018-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/linalg/distance_type.h>
+#include <raft/handle.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+#include <cstddef>
+
+namespace ML {
+namespace HDBSCAN {
+namespace Common {
+
+template <typename value_idx, typename value_t>
+class CondensedHierarchy {
+ public:
+  /**
+   * Constructs an empty condensed hierarchy object which requires
+   * condense() to be called in order to populate the state.
+   * @param handle_
+   * @param n_leaves_
+   */
+  CondensedHierarchy(const raft::handle_t &handle_, size_t n_leaves_);
+
+  /**
+   * Constructs a condensed hierarchy object with existing arrays
+   * which already contain a condensed hierarchy.
+   * @param handle_
+   * @param n_leaves_
+   * @param size_
+   * @param n_edges_
+   * @param parents_
+   * @param children_
+   * @param lambdas_
+   * @param sizes_
+   */
+  CondensedHierarchy(const raft::handle_t &handle_, size_t n_leaves_,
+                     int n_edges_, value_idx *parents_, value_idx *children_,
+                     value_t *lambdas_, value_idx *sizes_);
+
+  /**
+   * Constructs a condensed hierarchy object by moving
+   * rmm::device_uvector. Used to construct cluster trees
+   * @param handle_
+   * @param n_leaves_
+   * @param size_
+   * @param n_edges_
+   * @param parents_
+   * @param children_
+   * @param lambdas_
+   * @param sizes_
+   */
+  CondensedHierarchy(const raft::handle_t &handle_, size_t n_leaves_,
+                     int n_edges_, int n_clusters_,
+                     rmm::device_uvector<value_idx> &&parents_,
+                     rmm::device_uvector<value_idx> &&children_,
+                     rmm::device_uvector<value_t> &&lambdas_,
+                     rmm::device_uvector<value_idx> &&sizes_);
+  /**
+   * Populates the condensed hierarchy object with the output
```

Comment:
Could you also add a few words on what happens in condense apart from copying the data (transforming labels, sorting)?